sycl/include/CL/sycl/detail/context_impl.hpp

```diff
@@ -58,25 +59,19 @@ class context_impl {
   RT::PiContext &getHandleRef();
   const RT::PiContext &getHandleRef() const;
 
-  std::map<OSModuleHandle, RT::PiProgram> &getCachedPrograms() {
-    return m_CachedPrograms;
-  }
-  std::map<RT::PiProgram, std::map<string_class, RT::PiKernel>> &
-  getCachedKernels() {
-    return m_CachedKernels;
-  }
-
   std::shared_ptr<usm::USMDispatcher> getUSMDispatch() const;
+
+  KernelProgramCache &getKernelProgramCache() const;
+
 private:
   async_handler m_AsyncHandler;
   vector_class<device> m_Devices;
   RT::PiContext m_Context;
   platform m_Platform;
   bool m_OpenCLInterop;
   bool m_HostContext;
-  std::map<OSModuleHandle, RT::PiProgram> m_CachedPrograms;
-  std::map<RT::PiProgram, std::map<string_class, RT::PiKernel>> m_CachedKernels;
   std::shared_ptr<usm::USMDispatcher> m_USMDispatch;
+  std::unique_ptr<KernelProgramCache> m_KernelProgramCache;
```

Comment:
Why do you need a pointer? Better to just have it as a value.

Reply:
It's a trade-off for ability to return non-const ref from const method.
Another approach is to mark m_KernelProgramCache as `mutable`.
Don't know which way is better.

Reply:
So you have a const context_impl, but you need to access (modify) the cache? Why it is const then?

Reply:
The context is not changed, right? We say it to the user of `context_impl` by marking the method with `const` keyword. At the same time we provide a facility to change the cache. Cache is quite a supplementary thing to context. We may freely change the cache (e.g. clear it due to retention timeout) without changing the context.

Reply:
@romanovvlad please take a look.
If you do want to make a part of a const structure mutable, then it is better to use `mutable` keyword instead of a pointer.

Reply:
Employed `mutable` kw